third_party/amd/lib/TritonAMDGPUTransforms/inThreadTranspose.cpp

```diff
@@ -0,0 +1,213 @@
+  // `make_tensor_ptr`
+  SmallVector<Value, 4> newArgs;
+  for (auto operand : op->getOperands()) {
+    auto tensorType = dyn_cast<RankedTensorType>(operand.getType());
+    if (tensorType &&
+        !isa<triton::gpu::SharedEncodingAttr>(tensorType.getEncoding())) {
+      Type newType = getNewType(tensorType, encoding);
+      newArgs.push_back(builder.create<triton::gpu::ConvertLayoutOp>(
+          op->getLoc(), newType, operand));
+    } else {
+      newArgs.push_back(operand);
+    }
+  }
+
+  // Convert output types
+  SmallVector<Type, 4> newTypes;
+  for (auto t : op->getResultTypes()) {
+    bool isAsync = isa<triton::gpu::AsyncCopyGlobalToLocalOp>(op);
+    newTypes.push_back(isAsync ? t : getNewType(t, encoding));
+  }
+
+  // Construct new op with the new encoding
+  Operation *newOp = builder.create(op->getLoc(), op->getName().getIdentifier(),
+                                    newArgs, newTypes, op->getAttrs());
+
+  // Cast the results back to the original layout
+  for (size_t i = 0; i < op->getNumResults(); i++) {
+    Value newResult = newOp->getResult(i);
+    if (newTypes[i] != op->getResultTypes()[i]) {
+      newResult = builder.create<triton::gpu::ConvertLayoutOp>(
+          op->getLoc(), op->getResult(i).getType(), newResult);
+    }
+    op->getResult(i).replaceAllUsesWith(newResult);
+  }
+  op->erase();
+}
+
+SmallVector<Operation *> getLoadInsts(Operation *op) {
+  SmallVector<Operation *> ret;
+  // Deal with the case that convert_layout intakes a scf.if, etc.
+  auto v = op->getOperand(0);
+  auto prevOp = v.getDefiningOp();
+  if (isa<RegionBranchOpInterface>(prevOp)) {
+    LDBG("Dealing with scf blocks");
+    auto idx = cast<OpResult>(v).getResultNumber();
+    llvm::SmallVector<scf::YieldOp> yieldOps;
+    prevOp->walk([&](Operation *op) {
+      if (auto yieldOp = dyn_cast<scf::YieldOp>(op)) {
+        yieldOps.push_back(yieldOp);
+      }
+    });
+
+    for (auto yieldOp : yieldOps) {
+      auto maybeLoadOp = yieldOp.getOperand(idx).getDefiningOp();
+      if (isa<tt::LoadOp>(maybeLoadOp))
+        ret.push_back(maybeLoadOp);
+    }
+  } else {
+    // regular case
+    LDBG("Regular cases");
+    assert(isa<tt::LoadOp>(prevOp) &&
+           "only load->convert_layout->dot is supported.");
+    ret.push_back(prevOp);
+  }
+  return ret;
+}
+
+bool convertToThreadRaked(Value operand) {
+  auto opTensorTy = cast<RankedTensorType>(operand.getType());
+  auto opEnc = opTensorTy.getEncoding();
+  auto opDotOpEnc = dyn_cast<ttg::DotOperandEncodingAttr>(opEnc);
+  if (!opDotOpEnc)
+    return false;
+  if (!isa<ttg::AMDMfmaEncodingAttr>(opDotOpEnc.getParent())) {
+    LDBG("Operand's parent encoding is not MFMA");
+    return false;
+  }
+  auto cvtOperand = operand.getDefiningOp()->getOperand(0);
+  auto cvtOperandEnc =
+      cast<RankedTensorType>(cvtOperand.getType()).getEncoding();
+  auto blockedEnc = dyn_cast<ttg::BlockedEncodingAttr>(cvtOperandEnc);
```

Comment:
If `operand` is from dotOp.getA() or getB(), then operand.getDefiningOp() is the local_load. Can you double check the logic here?

Reply:
nvm. I realized this pass is introduced before shared memory is introduced.